### DISCUSSION TITLE
Internalization with TLDs

### DISCUSSION THREAD

**Discussion by shishkin17** [vikejs/vike#1453] [Discussion] [answered]
Are there any examples of how to do internalization with "domain.fr domain TLDs" that is mentioned on https://vike.dev/i18n?

**Comment by brillout** [comment]
Not that I'm aware of. Contribution welcome to create one.

I don't foresee any blockers, but let us know if you run into any.

**Reply by shishkin17** [reply]
Can you give in idea of what to use for that?
Example with `urlOriginal` is domain agnostic. I don't see how `origin` can be used with pre-rendering for different locales.

**Reply by lourot** [reply]
The requested domain name should be available in the incoming request. For example if you're using Express, it should be available as `req.hostname`. You can then inject that value in the `pageContext` via [pageContextInit](https://vike.dev/pageContext#custom). You can then pass `pageContext.hostname` to your [extractLocale()](https://vike.dev/i18n) implementation.

**Reply by shishkin17** [reply]
Will that approach work for pre-rendering on localhost (for future deployment on static host)?

And I can't fully understand how to modify `urlOriginal` (or something else) in `onPrerenderStart` in case of TLDs.
For case when you add paths like /locale/path its clear. But how should build structure look like in case of TLDs?

**Reply by lourot** [reply, marked as answer]
You're right, this approach won't work with pre-rendering for two reasons:

1. You then don't have the server glue code to read `req.hostname`
2. If `mydomain.fr/` and `mydomain.de/` both point to the exact same statically generated `index.html`, how can this `index.html` be possibly both French and German at the same time?

Point 2. illustrates IMHO that if you want to do pre-rendering with SSR in this case, you basically need to statically generate two entire and separate websites: one in French, one in German. And you then host them on two different say GitHub Pages.

Now I'm sure that's possible to generate these two websites from the same code base but that's not what is described in https://vike.dev/i18n . At this point I can't walk you through the solution off the top of my head. It's something I would have to explore myself. But I don't foresee any blocker with Vike.

**Reply by shishkin17** [reply]
Thank you for confirming my thoughts.
I think I'll try to dive deeper into such generation later.
